lib/api/CDataFrameTrainBoostedTreeClassifierRunner.cc

```diff
@@ -153,6 +153,20 @@ void CDataFrameTrainBoostedTreeClassifierRunner::writeOneRow(
         }
         writer.EndArray();
     }
+
+    if (this->topShapValues() > 0) {
+        auto shapColumnsRange{this->boostedTree().columnsHoldingShapValues()};
```

Comment:
At least in theory `nth_element` needs to guard against the case `topShapValues() > shapColumnsRange.size()` unless that happens elsewhere? (Also, we probably don't need to actually copy the size vector; we could have a begin and end or hold the vector and return by reference, at which point it is more efficient to have a heap of the top values.) We have a class which makes all this easy in the `maths` library; I think using this works out as clean in code. You'd need a factory method to take care of the type of the comparison operator, but something like should work:
```
template<typename LESS>
auto makeLargestShapAccumulator(std::size_t n, LESS less) {
    return maths::CBasicStatistics::COrderStatisticsHeap<std::size_t, LESS>{n, std::size_t{}, less);
}
...
    auto largestShapValues = makeLargestShapAccumulator(this->topShapValues(), [&row](std::size_t lhs, std::size_t rhs) {
                             return std::fabs(row[lhs]) > std::fabs(row[rhs]);
                         });
    largestShapValues.add(this->boostedTree().columnsHoldingShapValues());
    largestShapValues.sort();
    for (auto i : largestShapValues) {
       ...
    }
```

Reply:
Done. Great idea!